bitcoin/src/psbt/serialize.rs

```diff
@@ -18,26 +18,99 @@ use secp256k1::{self, XOnlyPublicKey};
 use crate::bip32::{ChildNumber, Fingerprint, KeySource};
 use crate::hashes::{hash160, ripemd160, sha256, sha256d, Hash};
 use crate::crypto::{ecdsa, schnorr};
-use crate::psbt;
+use crate::psbt::{self, Error, PartiallySignedTransaction};
 use crate::taproot::{TapBranchHash, TapLeafHash, ControlBlock, LeafVersion};
 use crate::crypto::key::PublicKey;
 
-use super::map::{TapTree, PsbtSighashType};
+use super::map::{Map, Input, Output, TapTree, PsbtSighashType};
 
 use crate::taproot::TaprootBuilder;
 /// A trait for serializing a value as raw data for insertion into PSBT
-/// key-value pairs.
+/// key-value maps.
 pub trait Serialize {
     /// Serialize a value as raw data.
     fn serialize(&self) -> Vec<u8>;
 }
 
-/// A trait for deserializing a value from raw data in PSBT key-value pairs.
+/// A trait for deserializing a value from raw data in PSBT key-value maps.
 pub trait Deserialize: Sized {
     /// Deserialize a value from raw data.
     fn deserialize(bytes: &[u8]) -> Result<Self, encode::Error>;
 }
 
+impl PartiallySignedTransaction {
+     /// Serialize a value as bytes in hex.
+     pub fn serialize_hex(&self) -> String {
+        bitcoin_hashes::hex::ToHex::to_hex(&self.serialize()[..])
+    }
+
+    /// Serialize a value as raw binary data.
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut buf: Vec<u8> = Vec::new();
+
+        buf.extend(b"psbt");
+    
+        buf.push(0xff_u8);
+
+        self.serialize_map(&mut buf).unwrap();
+
+        for i in &self.inputs {
+            i.consensus_encode(&mut buf).unwrap();
+        }
+
+        for i in &self.outputs {
+            i.consensus_encode(&mut buf).unwrap();
+        }
+
+        buf
+    }
+
+    /// Deserialize a value from raw binary data.
+    pub fn deserialize(bytes: &[u8]) -> Result<Self, encode::Error> {
+        let magic = &bytes[0..4];
```

Comment:
@DanGould, the fuzz test is failing here. You need to check if bytes is at least 4 len before indexing it.

Reply:
To reproduce fuzz failures, we you can look at the offending hex. "62" in this case https://github.com/rust-bitcoin/rust-bitcoin/actions/runs/3659817137/jobs/6186247711#step:6:110

and paste in `duplicate_crash` function of failing fuzz test (`deserialize_psbt` in this case). run `cargo test` to debug more.